prow/tide/gerrit.go

```diff
@@ -0,0 +1,324 @@
+var _ provider = (*GerritProvider)(nil)
+
+// GerritProvider implements provider, used by tide Controller for
+// interacting directly with Gerrit.
+//
+// Tide Controller should only use GerritProvider for communicating with Gerrit.
+type GerritProvider struct {
+	cfg config.Getter
+	gc  gerritClient
+
+	pjclientset ctrlruntimeclient.Client
+
+	cookiefilePath    string
+	tokenPathOverride string
+
+	*mergeChecker
+	logger *logrus.Entry
+}
+
+func newGerritProvider(
+	logger *logrus.Entry,
+	cfg config.Getter,
+	pjclientset ctrlruntimeclient.Client,
+	mergeChecker *mergeChecker,
+	cookiefilePath string,
+	tokenPathOverride string,
+) *GerritProvider {
+	gerritClient, err := client.NewClient(nil)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error creating gerrit client.")
+	}
+
+	gp := &GerritProvider{
+		logger:            logger,
+		cfg:               cfg,
+		pjclientset:       pjclientset,
+		gc:                gerritClient,
+		cookiefilePath:    cookiefilePath,
+		tokenPathOverride: tokenPathOverride,
+		mergeChecker:      mergeChecker,
+	}
+
+	gp.applyGlobalConfig(cfg, gerritClient, cookiefilePath, tokenPathOverride)
+
+	return gp
+}
+
+func (p *GerritProvider) applyGlobalConfig(cfg config.Getter, gerritClient *client.Client, cookiefilePath, tokenPathOverride string) {
+	p.applyGlobalConfigOnce(cfg, gerritClient, cookiefilePath, tokenPathOverride)
+
+	go func() {
+		for {
+			p.applyGlobalConfigOnce(cfg, gerritClient, cookiefilePath, tokenPathOverride)
+			// No need to spin constantly, give it a break. It's ok that config change has one second delay.
+			time.Sleep(time.Second)
+		}
+	}()
+}
+
+func (p *GerritProvider) applyGlobalConfigOnce(cfg config.Getter, gerritClient *client.Client, cookiefilePath, tokenPathOverride string) {
+	gerritCfg := cfg().Tide.Gerrit
+	if gerritCfg == nil {
+		return
+	}
+
+	if err := gerritClient.UpdateClients(cfg().Gerrit.OrgReposConfig.AllRepos()); err != nil {
+		logrus.WithError(err).Error("Updating clients.")
+	}
+	// Authenticate creates a goroutine for rotating token secrets when called the first
+	// time, afterwards it only authenticate once.
+	gerritClient.Authenticate(cookiefilePath, tokenPathOverride)
+}
+
+// Query returns all PRs from configured gerrit org/repos.
+func (p *GerritProvider) Query() (map[string]CodeReviewCommon, error) {
+	// lastUpdate is used by gerrit adapter for achieving incremental query. In
+	// tide case we want to get everything so use default time.Time, which
+	// should be 1970,1,1.
+	var lastUpdate time.Time
+
+	res := make(map[string]CodeReviewCommon)
+	// This is querying serially, which would safely guard against quota issues.
+	// TODO(chaodai): parallize this to boot the performance.
+	for instance, projs := range p.cfg().Tide.Gerrit.Queries.AllRepos() {
+		for projName := range projs {
+			changes, err := p.gc.QueryChangesForProject(instance, projName, lastUpdate, p.cfg().Gerrit.RateLimit, []string{gerritDefaultQueryParam})
```

Comment:
The comment for this function indicates the caller must manage the last sync state, but I don't see that done here. I guess this is not applicable to Tide though since we want to consider all open changes? If so the comment seems more applicable to the gerrit adapter rather than the gerrit client so we should probably move it.

Reply:
yeah you're right about that this is not applicable to tide. There are a couple more `Query*` functions from the `client` that do manage the last sync time struct, so my feeling is that the comment applies to client method. However I have expanded the comment to make it more clear.